File: app/src/main/java/com/imangazalievm/bubbble/presentation/mvp/shotdetails/ShotDetailsPresenter.java
package com.imangazalievm.bubbble.presentation.mvp.shotdetails;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.imangazalievm.bubbble.domain.global.exceptions.NoNetworkException;
import com.imangazalievm.bubbble.domain.global.models.Comment;
import com.imangazalievm.bubbble.domain.global.models.Shot;
import com.imangazalievm.bubbble.domain.global.models.ShotCommentsRequestParams;
import com.imangazalievm.bubbble.domain.shotdetails.ShotDetailsInteractor;
import com.imangazalievm.bubbble.presentation.mvp.global.SchedulersProvider;
import com.imangazalievm.bubbble.presentation.mvp.global.permissions.Permission;
import com.imangazalievm.bubbble.presentation.mvp.global.permissions.PermissionsManager;
import com.imangazalievm.bubbble.presentation.mvp.global.permissions.PermissionsManagerHolder;
import com.imangazalievm.bubbble.presentation.utils.DebugUtils;

import java.util.List;

import javax.inject.Inject;

@InjectViewState
public class ShotDetailsPresenter extends MvpPresenter<ShotDetailsView> {

    private static final int COMMENTS_PAGE_SIZE = 20;

    private ShotDetailsInteractor shotDetailsInteractor;
    private SchedulersProvider schedulersProvider;
    private PermissionsManagerHolder permissionsManagerHolder;
    private long shotId;
    private Shot shot;
    private int currentMaxCommentsPage = 1;
    private boolean isCommentsLoading = false;

    @Inject
    public ShotDetailsPresenter(
            ShotDetailsInteractor shotDetailsInteractor,
            SchedulersProvider schedulersProvider,
            long shotId
    ) {
        this.shotDetailsInteractor = shotDetailsInteractor;
        this.permissionsManagerHolder = new PermissionsManagerHolder();
        this.schedulersProvider = schedulersProvider;
        this.shotId = shotId;
    }

    public void setPermissionsManager(PermissionsManager permissionsManager) {
        permissionsManagerHolder.setPermissionsManager(permissionsManager);
    }

    public void removePermissionsManager() {
        permissionsManagerHolder.removePermissionsManager();
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();

        loadShot();
    }

    private void loadShot() {
        getViewState().showLoadingProgress();
        shotDetailsInteractor.getShot(shotId)
                .observeOn(schedulersProvider.ui())
                .subscribe(this::onShotLoaded, this::onShotLoadError);
    }

    private void onShotLoaded(Shot shot) {
        this.shot = shot;
        getViewState().hideLoadingProgress();
        getViewState().showShot(shot);

        if (shot.getCommentsCount() > 0) {
            loadMoreComments(1);
        } else {
            getViewState().showNoComments();
        }
    }

    private void onShotLoadError(Throwable throwable) {
        if (throwable instanceof NoNetworkException) {
            getViewState().hideLoadingProgress();
            getViewState().showNoNetworkLayout();
        } else {
            DebugUtils.showDebugErrorMessage(throwable);
        }
    }

    public void onImageLoadError() {
        getViewState().hideImageLoadingProgress();
    }

    public void onImageLoadSuccess() {
        getViewState().hideImageLoadingProgress();
    }

    public void retryLoading() {
        getViewState().hideNoNetworkLayout();
        loadShot();
    }

    private void loadMoreComments(int page) {
        isCommentsLoading = true;
        ShotCommentsRequestParams shotCommentsRequestParams = new ShotCommentsRequestParams(shotId, page, COMMENTS_PAGE_SIZE);
        shotDetailsInteractor.getShotComments(shotCommentsRequestParams)
                .observeOn(schedulersProvider.ui())
                .subscribe(this::onCommentsLoaded, DebugUtils::showDebugErrorMessage);
    }

    private void onCommentsLoaded(List<Comment> newComments) {
        isCommentsLoading = false;
        getViewState().hideCommentsLoadingProgress();
        getViewState().showNewComments(newComments);
    }


    public void onLoadMoreCommentsRequest() {
        if (isCommentsLoading) {
            return;
        }
        currentMaxCommentsPage++;
        loadMoreComments(currentMaxCommentsPage);
    }

    public void onImageClicked() {
        getViewState().openShotImageScreen(shot);
    }

    public void onLikeShotClicked() {

    }

    public void onShareShotClicked() {
        getViewState().showShotSharing(shot.getTitle(), shot.getHtmlUrl());
    }

    public void onDownloadImageClicked() {
        if (permissionsManagerHolder.checkPermissionGranted(Permission.READ_EXTERNAL_STORAGE)) {
            saveShotImage();
        } else {
            permissionsManagerHolder.requestPermission(Permission.READ_EXTERNAL_STORAGE, permissionResult -> {
                if (permissionResult.granted) {
                    saveShotImage();
                } else if (permissionResult.shouldShowRequestPermissionRationale) {
                    getViewState().showStorageAccessRationaleMessage();
                } else {
                    getViewState().showAllowStorageAccessMessage();
                }
            });
        }
    }

    public void onAppSettingsButtonClicked() {
        getViewState().openAppSettingsScreen();
    }

    private void saveShotImage() {
        shotDetailsInteractor.saveImage(shot.getImages().best())
                .observeOn(schedulersProvider.ui())
                .subscribe(() -> getViewState().showImageSavedMessage(), DebugUtils::showDebugErrorMessage);
    }

    public void onOpenShotInBrowserClicked() {
        getViewState().openInBrowser(shot.getHtmlUrl());
    }

    public void onShotAuthorProfileClicked() {
        getViewState().openUserProfileScreen(shot.getUser().getId());
    }

    public void onCommentAuthorClick(long userId) {
        getViewState().openUserProfileScreen(userId);
    }

    public void onLinkClicked(String url) {
        getViewState().openInBrowser(url);
    }

    public void onTagClicked(String tag) {

    }

}

File: app/src/main/java/com/imangazalievm/bubbble/presentation/shotdetails/ShotDetailsPresenter.kt
package com.imangazalievm.bubbble.presentation.shotdetails

import com.afollestad.assent.Permission
import com.arellomobile.mvp.InjectViewState
import com.imangazalievm.bubbble.domain.global.models.Shot
import com.imangazalievm.bubbble.domain.global.models.ShotCommentsRequestParams
import com.imangazalievm.bubbble.domain.shotdetails.ShotDetailsInteractor
import com.imangazalievm.bubbble.presentation.global.mvp.BasePresenter
import com.imangazalievm.bubbble.presentation.global.permissions.PermissionsManager
import javax.inject.Inject

@InjectViewState
class ShotDetailsPresenter @Inject constructor(
    private val shotDetailsInteractor: ShotDetailsInteractor,
    private val permissionsManager: PermissionsManager,
    private val shotId: Long
) : BasePresenter<ShotDetailsView>() {

    private lateinit var shot: Shot
    private val isShotLoaded: Boolean
        get() = ::shot.isInitialized
    private var currentMaxCommentsPage = 1
    private var isCommentsLoading = false

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()

        loadShot()
    }

    private fun loadShot() = launchSafe {
        try {
            viewState.showLoadingProgress()
            this@ShotDetailsPresenter.shot = shotDetailsInteractor.getShot(shotId)
            viewState.hideLoadingProgress()
            viewState.showShot(shot)
            if (shot.commentsCount > 0) {
                loadMoreComments(1)
            } else {
                viewState.showNoComments()
            }
        } finally {
            viewState.hideLoadingProgress()
            viewState.showNoNetworkLayout()
        }
    }

    fun onImageLoadError() {
        viewState.hideImageLoadingProgress()
    }

    fun onImageLoadSuccess() {
        viewState.hideImageLoadingProgress()
    }

    fun retryLoading() {
        viewState.hideNoNetworkLayout()
        loadShot()
    }

    private fun loadMoreComments(page: Int) = launchSafe {
        isCommentsLoading = true
        val shotCommentsRequestParams = ShotCommentsRequestParams(shotId, page, COMMENTS_PAGE_SIZE)
        val newComments = try {
            shotDetailsInteractor.getShotComments(shotCommentsRequestParams)
        } finally {
            isCommentsLoading = false
        }
        viewState.hideCommentsLoadingProgress()
        viewState.showNewComments(newComments)
    }

    fun onLoadMoreCommentsRequest() {
        if (isCommentsLoading) {
            return
        }
        currentMaxCommentsPage++
        loadMoreComments(currentMaxCommentsPage)
    }

    fun onImageClicked() {
        viewState.openShotImageScreen(shot)
    }

    fun onLikeShotClicked() {}

    fun onShareShotClicked() {
        if (!isShotLoaded) return
        viewState.showShotSharing(shot.title, shot.htmlUrl)
    }

    fun onDownloadImageClicked() = launchSafe {
        if (!isShotLoaded) return@launchSafe
        if (permissionsManager.isGranted(Permission.READ_EXTERNAL_STORAGE)) {
            saveShotImage()
        } else {
            val result = permissionsManager.requestPermission(Permission.READ_EXTERNAL_STORAGE)
            when {
                result.isGranted -> saveShotImage()
                result.isBlockedFromAsking -> viewState.showStorageAccessRationaleMessage()
                else -> viewState.showAllowStorageAccessMessage()
            }
        }
    }

    fun onAppSettingsButtonClicked() {
        viewState.openAppSettingsScreen()
    }

    private fun saveShotImage() = launchSafe {
        if (!isShotLoaded) return@launchSafe
        shotDetailsInteractor.saveImage(shot.images.best())
        viewState.showImageSavedMessage()
    }

    fun onOpenShotInBrowserClicked() {
        viewState.openInBrowser(shot.htmlUrl)
    }

    fun onShotAuthorProfileClicked() {
        viewState.openUserProfileScreen(shot.user.id)
    }

    fun onCommentAuthorClick(userId: Long) {
        viewState.openUserProfileScreen(userId)
    }

    fun onLinkClicked(url: String) {
        viewState.openInBrowser(url)
    }

    fun onTagClicked(tag: String) {}

    companion object {
        private const val COMMENTS_PAGE_SIZE = 20
    }

}